### ISSUE TITLE
Duplicated Wandb runs with multiple gpus 

### ISSUE THREAD

**Issue by dantetemplar** [ultralytics/ultralytics#16704]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Multi-GPU

### Bug

For one actual training run multiple Wandb runs are spawned when I use multiple gpus. I guess that problem because of use of `subprocess.run`, so no global `wandb.run` will be propagated to new process.

![image](https://github.com/user-attachments/assets/2fcc2c32-fd13-4ad7-a036-d68bc6ba45d7)
![image](https://github.com/user-attachments/assets/069c60ab-96d3-4ded-8556-7ea76b5388aa)

I suggest to propagate wandb run's project, name, and id to subprocess through [Environment](https://docs.wandb.ai/guides/track/environment-variables/) or ddp file (hard way), that should fix the issue.

### Environment

```
Ultralytics 8.3.5 🚀 Python-3.10.14 torch-2.4.0 CUDA:0 (Tesla T4, 15095MiB)
Setup complete ✅ (4 CPUs, 31.4 GB RAM, 5934.0/8062.4 GB disk)
```

Kaggle with GPU T4 x 2

### Minimal Reproducible Example

https://www.kaggle.com/code/dantetemplarwhowas/notebook915070977e

### Additional

It's hard for me to debug actual behavior because I have only one gpu on my local machine. Problem appears in Kaggle with GPU T4 x 2 accelerator.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @dantetemplar, thank you for your interest in Ultralytics 🚀!

We recommend checking out the [Docs](https://docs.ultralytics.com) where you'll find comprehensive resources and examples. For new users, our [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) guides can be quite helpful.

It looks like you've encountered a 🐛 Bug related to Multi-GPU and Wandb runs. To help us debug, please ensure you've provided a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/), which you've linked—thank you! An Ultralytics engineer will review it soon to assist further.

For real-time discussions, join our [Discord](https://ultralytics.com/discord) 🎧. If you prefer more in-depth threads, visit [Discourse](https://community.ultralytics.com). You can also share insights on our [Subreddit](https://reddit.com/r/ultralytics) with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package to ensure your issue isn't already resolved:

```bash
pip install -U ultralytics
```

## Verified Environments

YOLO can run in these verified environments:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**, **Amazon**, **Docker**: See respective [Quickstart Guides](https://docs.ultralytics.com/yolov5/environments/)

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If the badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing, verifying correct operations across platforms.

An engineer will be with you shortly to delve deeper. Thank you for your patience and contribution!

**Comment by Y-T-G** [comment]
`wandb` is integrated natively into ultralytics. You should be using the integrated functionality.

https://github.com/ultralytics/ultralytics/blob/fe61f9d54a69c837a09c4f18668aecc169556e96/ultralytics/utils/callbacks/wb.py#L110-L112

**Comment by dantetemplar** [comment]
> `wandb` is integrated natively into ultralytics. You should be using the integrated functionality.
> 
> https://github.com/ultralytics/ultralytics/blob/fe61f9d54a69c837a09c4f18668aecc169556e96/ultralytics/utils/callbacks/wb.py#L110-L112

First of all, that will not fix the issue - two runs will be spawned. Also, such usage lacks parameters (most important one - entity). Moreover, ultralytics docs suggests multiple ways to integrate with wandb.

**Comment by Y-T-G** [comment]
> First of all, that will not fix the issue - two runs will be spawned. 

Using the native implementation doesn't spawn two sessions. It only runs the callback on the GPU with RANK 0, and spawns a single session.


> Also, such usage lacks parameters (most important one - entity). Moreover, ultralytics docs suggests [multiple ways](https://docs.ultralytics.com/integrations/weights-biases/#how-do-i-train-a-yolo11-model-using-weights-biases) to integrate with wandb.


The tutorial wouldn't work with multi-GPU since neither multi-GPU training neither respects custom callback, nor knows about `wandb` session initiated in a different process. The native implementation is the recommended way for multi-GPU training.

**Comment by Y-T-G** [comment]
> First of all, that will not fix the issue - two runs will be spawned. 

Using the native implementation (i.e., not explicitly initiating the wandb but letting ultralytics initiate it) doesn't spawn two sessions. It only runs the callback on the GPU with RANK 0, and spawns a single session.


> Also, such usage lacks parameters (most important one - entity). Moreover, ultralytics docs suggests [multiple ways](https://docs.ultralytics.com/integrations/weights-biases/#how-do-i-train-a-yolo11-model-using-weights-biases) to integrate with wandb.


The tutorial wouldn't work with multi-GPU since multi-GPU training neither respects custom callback, nor knows about `wandb` session initiated in a different process. The native implementation is the recommended way for multi-GPU training.

Setting the environment variable after `wand.init` and before `model.train()` also seems to work. So you can use that if you want to explicitly perform `wandb.init`:
```
import os
os.environ["WANDB_RUN_ID"] = wandb.run.id
```

**Comment by dantetemplar** [comment]
Also need to pass Entity, may be project too. 

Now two runs are spawned: one in team(entity), other in personal account.

![image](https://github.com/user-attachments/assets/f6efd88b-3201-4955-8237-8aec4beb7e42)
![image](https://github.com/user-attachments/assets/275f19ab-14a2-4c90-9295-43c7651a5a66)

**Comment by Y-T-G** [comment]
You can set the environment variables for that too 

`WANDB_PROJECT` and `WANDB_ENTITY`

**Comment by dantetemplar** [comment]
Suggest such kind of example in docs:
```python
import os
import wandb

from ultralytics import YOLO
from wandb.integration.ultralytics import add_wandb_callback

wandb.login()

model = YOLO(model="yolo11s.pt", task="detect")

add_wandb_callback(
    model,
    enable_model_checkpointing=True,
    enable_train_validation_logging=False,
    enable_validation_logging=False,
    enable_prediction_logging=False,
)

with wandb.init(
    entity="entity",
    project="project",
    name="yolo11s",
    notes="YOLO11s on VisDrone dataset",
    job_type="training",
    tags=["visdrone", "yolo11s"],
    reinit=True,
) as run:
    os.environ["WANDB_ENTITY"] = run.entity
    os.environ["WANDB_PROJECT"] = run.project
    os.environ["WANDB_RUN_ID"] = run.id
    
    results = model.train(
        project=run.project,
        name=run.name,
        data="VisDrone_people.yaml",
        cache="ram",
        device=[0, 1],
        batch=64
    )
```

**Comment by glenn-jocher** [comment]
Thank you for the suggestion. We appreciate your input and will consider it for future documentation updates.

**Comment by glenn-jocher** [comment]
@dantetemplar actually feel free to submit a PR for this. We just updated the page but no harm in adding an additional example, though `project` and `name` automatically get passed to W&B so nothing to really add there. I've never done anything except `wandb login` and then train a model myself. That's all 99% of people need.

**Comment by dantetemplar** [comment]
> @dantetemplar actually feel free to submit a PR for this. We just updated the page but no harm in adding an additional example, though `project` and `name` automatically get passed to W&B so nothing to really add there. I've never done anything except `wandb login` and then train a model myself. That's all 99% of people need.

Okay, I will try to contribute after I fix all issues. Now there is another problem.

On table view there are no any metrics for multiple gpus (the first one - training on two gpus, others - on one gpu)
![image](https://github.com/user-attachments/assets/0518c9a5-c23f-4eab-931d-1bac8f515903)

But the metrics exist for run actually.
![image](https://github.com/user-attachments/assets/23da3537-3e01-42d2-bee3-510964b6e523)

**Comment by dantetemplar** [comment]
I guess there are more problems regarding callbacks with multiple-gpu setup

**Comment by dantetemplar** [comment]
@glenn-jocher what do you prefer Wandb or ClearML? And why?

**Comment by glenn-jocher** [comment]
DDP training should log correctly. If not then there's a possible bug.

**Comment by dantetemplar** [comment]
> DDP training should log correctly. If not then there's a possible bug.

I see RANK environment guard for [hub callbacks](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/callbacks/hub.py#L12) but not [for wandb](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/callbacks/wb.py#L112). I guess there is need for such guard and some changes for work with wandb (at least, add to trainer attribute wandb_run). 

Also I have question: What the difference between RANK (global rank) and LOCAL_RANK?

**Comment by dantetemplar** [comment]
I think a more obvious and understandable behavior for library users will be enabling wandb integration ONLY IF wandb.run currently exists. That is, the only option will be wandb.init, and then model.train. Excluding the behavior: model.train itself creates run in the YOLO8 project with a random id, and so on.

**Comment by glenn-jocher** [comment]
Thank you for your suggestion. We appreciate your input, but the current behavior is designed to ensure seamless integration for most users. Feel free to submit a PR if you have a specific implementation in mind.
